core/common/src/ContourDisplay.ts

```diff
@@ -0,0 +1,263 @@
+
+      if (5 !== this.majorIntervalCount)
+        props.majorIntervalCount = this.majorIntervalCount;
+
+      return props;
+    }
+  }
+
+  /** JSON representation of the description of how contours should appear for a particular set of subcategories. */
+  export interface TerrainProps {
+    /** See [[Terrain.contourDef]]. */
+    contourDef?: ContourProps;
+    /** See [[Terrain.subCategories]]. */
+    subCategories?: Id64String[];
+  }
+
+  /** Contains a description of how contours should appear for a particular set of subcategories. A terrain is an organizational concept which associates a contour appearance with a list of subcategories within an iModel.
+   * @see [[ContourDisplay.Contour]]
+   */
+  export class Terrain {
+    /** A [[ContourDisplay.Contour]] object describing how the contours for this terrain should appear. Defaults to an instantation of [[ContourDisplay.Contour]] using all of its own default properties. */
+    public readonly contourDef: Contour;
+    /** List of subcategory IDs to which this terrain styling will be applied. */
+    public readonly subCategories: Id64String[];
+
+    public equals(other: Terrain | undefined): boolean {
+      if (this === undefined && other === undefined)
+        return true;
+      if (this === undefined || other === undefined)
+        return false;
+      if (!this.contourDef.equals(other.contourDef))
+        return false;
+      if (this.subCategories.length !== other.subCategories.length)
+        return false;
+      for (const subCategory of this.subCategories) {
+        const match = other.subCategories.find((element) => element === subCategory);
+        if (!match)
+          return false;
+      }
+      return true;
+    }
+
+    private constructor(json?: TerrainProps) {
+      if (undefined === json) {
+        this.contourDef = Contour.fromJSON({});
+        this.subCategories = [];
+      } else {
+        this.contourDef = json.contourDef ? Contour.fromJSON(json.contourDef) : Contour.fromJSON({});
+        this.subCategories = json.subCategories ? [...json.subCategories] : [];
+      }
+    }
+
+    public static fromJSON(json?: TerrainProps) {
+      return json ? new Terrain(json) : new Terrain({});
+    }
+
+    public toJSON(): TerrainProps {
+      const props: TerrainProps = {};
+
+      if (!this.contourDef.equals(Contour.defaults))
+        props.contourDef = this.contourDef.toJSON();
+
+      if (this.subCategories.length > 0)
+        props.subCategories = [...this.subCategories];
+
+      return props;
+    }
+  }
+
+  /** JSON representation of the contour display setup of a [[DisplayStyle3d]]. */
+  export interface SettingsProps {
+    /** See [[ContourDisplay.terrains]]. */
+    terrains?: (TerrainProps | undefined)[];
+  }
+
+  /** The contour display setup of a [[DisplayStyle3d]].
+   * Contour display allows a user to apply specific contour line renderings to subcategories within a scene.
+   */
+  export class Settings {
```

Comment:
Consider adding a `ViewFlags.contourLines` so that 1. people can toggle them on and off without destroying and recreating the contour settings and 2. people can create a `GraphicBranch` to contain graphics that shouldn't have contour lines applied to them. See `ViewFlags.thematicDisplay` for comparison.

Reply:
I added a `displayContours` toggle on the display settings.

Reply:
In that case you will need to raise another event when that flag changes.
Seems simpler to put the flag in the ContourDisplay object.

Reply:
That would mean the ContourDisplay object will need to destroy and recreate itself in order to change the flag, which this change was hoping to avoid.  That is, if we make the flag living on ContourDisplay follow the pattern it currently does --- exposing access to everything only through fromJSON() and toJSON().

Are you proposing the flag on the ContourDisplay object be accessible without going through those methods?

Reply:
ContourDisplay and the objects it contains are all immutable, there is no need to destroy any of them. Look at `Environment.withDisplay` for the equivalent.